File: IOS/Sensors/Accelerometer.m
// Created by Michael Simms on 7/14/12.
// Copyright (c) 2012 Michael J. Simms. All rights reserved.

// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

#import "Accelerometer.h"

@implementation Accelerometer

- (id)init
{
	self = [super init];
	if (self != nil)
	{
		self->motionManager = [[CMMotionManager alloc] init];
	}
	return self;
}

#pragma mark Sensor methods

- (void)enteredBackground
{	
}

- (void)enteredForeground
{
}

- (SensorType)sensorType
{
	return SENSOR_TYPE_ACCELEROMETER;
}

- (void)startUpdates
{
	if (self->motionManager && [self->motionManager isAccelerometerAvailable])
	{
#if TARGET_OS_WATCH
		[self->motionManager setAccelerometerUpdateInterval:0.2];
#else
		[self->motionManager setAccelerometerUpdateInterval:0.1];
#endif
		[self->motionManager startAccelerometerUpdatesToQueue:[NSOperationQueue mainQueue] withHandler:^(CMAccelerometerData* accelerometerData, NSError* error)
		 {
			 NSDate* now = [NSDate date];
			 uint64_t theTimeMs = (uint64_t)([now timeIntervalSince1970] * (double)1000.0);

			 CMAccelerometerData* rawData = [self->motionManager accelerometerData];
			 NSDictionary* accelData = [[NSDictionary alloc] initWithObjectsAndKeys:
										[NSNumber numberWithDouble:rawData.acceleration.x],@KEY_NAME_ACCEL_X,
										[NSNumber numberWithDouble:rawData.acceleration.y],@KEY_NAME_ACCEL_Y,
										[NSNumber numberWithDouble:rawData.acceleration.z],@KEY_NAME_ACCEL_Z,
										[NSNumber numberWithLongLong:theTimeMs], @KEY_NAME_ACCELEROMETER_TIMESTAMP_MS,
										nil];
			 [[NSNotificationCenter defaultCenter] postNotificationName:@NOTIFICATION_NAME_ACCELEROMETER object:accelData];
		 }];
	}
}

- (void)stopUpdates
{
	if (self->motionManager && [self->motionManager isAccelerometerAvailable])
	{
		[self->motionManager stopAccelerometerUpdates];
	}
}

- (void)update
{
}

@end

File: IOS/Sensors/Accelerometer.swift
//
//  Accelerometer.swift
//  Created by Michael Simms on 10/3/22.
//

import Foundation
import CoreMotion

// Subscribe to the notification with this name to receive updates.
let NOTIFICATION_NAME_ACCELEROMETER = "ALAccelerometerUpdated"

// Keys for the dictionary associated with the notification.
let KEY_NAME_ACCEL_X = AXIS_NAME_X
let KEY_NAME_ACCEL_Y = AXIS_NAME_Y
let KEY_NAME_ACCEL_Z = AXIS_NAME_Z
let KEY_NAME_ACCELEROMETER_TIMESTAMP_MS = "Time"

class Accelerometer {
	var motionManager: CMMotionManager = CMMotionManager()

	func start() {
		if self.motionManager.isAccelerometerAvailable {
#if TARGET_OS_WATCH
			self.motionManager.accelerometerUpdateInterval = 0.2
#else
			self.motionManager.accelerometerUpdateInterval = 0.1
#endif

			self.motionManager.startAccelerometerUpdates(to: OperationQueue.main, withHandler: { accelerometerData, error in
				let rawData = self.motionManager.accelerometerData
				let theTimeMs = Date().timeIntervalSince1970 * 1000
				var accelData: Dictionary<String, Double> = [:]
				accelData[KEY_NAME_ACCEL_X] = rawData?.acceleration.x
				accelData[KEY_NAME_ACCEL_Y] = rawData?.acceleration.y
				accelData[KEY_NAME_ACCEL_Z] = rawData?.acceleration.z
				accelData[KEY_NAME_ACCELEROMETER_TIMESTAMP_MS] = Double(theTimeMs)
				let notification = Notification(name: Notification.Name(rawValue: NOTIFICATION_NAME_ACCELEROMETER), object: accelData)
				NotificationCenter.default.post(notification)
			})
		}
	}

	func stop() {
		if self.motionManager.isAccelerometerAvailable {
			self.motionManager.stopAccelerometerUpdates()
		}
	}
}
